src/string/extglobmatch.c

```diff
@@ -0,0 +1,278 @@
+                    .op = p
+                };
+                pa++;
+                alter = 0;
+                atoms = 0;
+                p = *pat++;
+                break;
+
+            case '(':
+                printError(p);
+                goto error;
+
+            case ')': {
+                    glob *tmp, *a;
+                    if (pa == paren || !atoms) {
+                        printError(p);
+                        goto error;
+                    }
+                    --pa;
+                    catAll();
+                    a = popFrag();
+                    tmp = handleOperator(pa->op, a);
+                    freeGlob(a);
+                    if(!tmp){
+                        goto error;
+                    }
+
+                    pushFrag(tmp);
+
+                    for (; alter > 0; --alter) {
+                        glob *b = popFrag();
+                        glob *a = popFrag();
+                        if (pa->op == '!')
+                            pushFrag(And(a, b));
+                        else
+                            pushFrag(Or(a, b));
+                        --atoms;
+                        freeGlob(a);
+                        freeGlob(b);
+                    }
+
+                    atoms = pa->atoms;
+                    alter = pa->alter;
+                    atoms++;
+                }
+                break;
+
+            case '[': {
+                    char *strPat;
+                    char const *end;
+                    catOne();
+                    end = strchr(pat, ']');
+                    if (!end) {
+                        goto character;
+                    }
+                    strPat = strdup(pat);
+                    strPat[end-pat]='\0';
+                    pushFrag(CSet(strPat));
+                    ++atoms;
+                    free(strPat);
+                    pat = end + 1;
+                    break;
+                }
+
+            case '|': {
+                    glob *a;
+                    glob *tmp;
+                    if (atoms == 0 || pa == paren) {
+                        printError(p);
+                        goto error;
+                    }
+                    catAll();
+                    a = popFrag();
+                    tmp = handleOperator((pa-1)->op, a);
+                    freeGlob(a);
+                    if(!tmp) {
+                        goto error;
+                    }
+                    pushFrag(tmp);
+
+                    ++alter;
+                }
+                break;
+
+            case '\\':
+                p = *pat++;
+                /* FALLTHROUGH */
+            default: /* literal character */
+character:     catOne();
+               pushFrag(Char(p));
+               ++atoms;
+        }
+
+    }
+    if (pa != paren) {
+        fprintf(stderr,"pattern error: Mismatching parenthesis\n");
+        goto error;
+    }
+
+    catAll();
+    for (; alter > 0; --alter) {
```

Comment:
It looks like this pattern was here to support compound patterns with a mismatching parenthesis, as the only moment where alter would be greater than 0 is because `')'` was not hit. This is already handled L152-L162, so I would drop this loop altogether.